### DISCUSSION TITLE
How do I upload a file?

### DISCUSSION THREAD

**Discussion by pengqian089** [reactiveui/refit#1545] [Q&A] [answered]
Using HttpClient, it should look like this:

``` csharp
private readonly HttpClient _httpClient;
ctor(HttpClient httpClient)
{
         _httpClient = httpClient;
}

async Task UploadAsync()
{
	var request = new HttpRequestMessage(HttpMethod.Put, "/images/2023/temp.jpg");
    // file content, byte[] or Stream
    byte[] buffer = null; // Todo
    Stream stream = null; // Todo

    request.Content = new ByteArrayContent(buffer);
    // or
    request.Content = new StreamContent(stream);

	var response = await _httpClient.SendAsync(request);
}
```

So how should Refit be used?

**Comment by glennawatson** [comment, marked as answer]
Use a `Multipart` method with one of the part wrapper types. Refit handles building the multipart content for you.

```csharp
public interface IUploadApi
{
    [Multipart]
    [Put("/images/2023/temp.jpg")]
    Task Upload([AliasAs("file")] StreamPart stream);

    // or for a byte[]
    [Multipart]
    [Put("/images/2023/temp.jpg")]
    Task Upload([AliasAs("file")] ByteArrayPart bytes);
}
```

Call it:

```csharp
await api.Upload(new StreamPart(stream, "temp.jpg", "image/jpeg"));
// or
await api.Upload(new ByteArrayPart(buffer, "temp.jpg", "image/jpeg"));
```

If you do NOT want multipart and instead want the raw body to be the file bytes/stream (matching your `ByteArrayContent` / `StreamContent` example), pass a `Stream` or `byte[]` as the body without `[Multipart]`:

```csharp
[Put("/images/2023/temp.jpg")]
Task Upload([Body] Stream content);
```

- A `Stream` body is sent via `StreamContent`; a `byte[]` body via `ByteArrayContent`.
- Wrapper types available for multipart parts: `StreamPart`, `ByteArrayPart`, `FileInfoPart`.

See the README "Multipart uploads" section for the full list and options.
